src/Passes/docs/base-profile-transformations.md

```diff
@@ -0,0 +1,233 @@
+  %q = phi i64 [ 8, %entry ], [ %4, %exiting__1 ]
+  %0 = icmp sle i64 %q, 10
+  br i1 %0, label %body__1, label %exit__1
+
+body__1:                                          ; preds = %header__1
+  %1 = sub i64 %q, 8
+  %2 = call i8* @__quantum__rt__array_get_element_ptr_1d(%Array* %qs, i64 %1)
+  %3 = bitcast i8* %2 to %Qubit**
+  %qubit = load %Qubit*, %Qubit** %3, align 8
+  call void @__quantum__qis__x__body(%Qubit* %qubit)
+  br label %exiting__1
+
+exiting__1:                                       ; preds = %body__1
+  %4 = add i64 %q, 1
+  br label %header__1
+
+exit__1:                                          ; preds = %header__1
+  call void @__quantum__rt__array_update_alias_count(%Array* %qs, i32 -1)
+  call void @__quantum__rt__qubit_release_array(%Array* %qs)
+  ret void
+}
+```
+
+After applying the our [demo profile transformation](https://github.com/troelsfr/qsharp-compiler/tree/feature/profile-pass/src/Passes/examples/QubitAllocationAnalysis), the QIR is reduced to:
+
+```
+define void @Feasibility__QubitMapping__Interop() local_unnamed_addr #0 {
+entry:
+  call void @__quantum__qis__x__body(%Qubit* null)
+  call void @__quantum__qis__x__body(%Qubit* nonnull inttoptr (i64 1 to %Qubit*))
+  call void @__quantum__qis__x__body(%Qubit* nonnull inttoptr (i64 2 to %Qubit*))
+  ret void
+}
+```
+
+We note that we successfully have eliminated loops, arithmetic operations, dynamic qubit allocation and alias counting - all operations which are not supported by the base profile.
+
+## Goal
+
+We envision the tool to work as a standalone commandline tool which can either validate or generate a QIR in accordance with a given profile. To validate, one would run:
+
+```language
+qat -p profile.yaml --validate unvalidated-qir.ll
+```
+
+In a similar fashion, generation is performed by adding `--generate` to the commandline:
+
+```language
+qat -p profile.yaml --generate qir.ll > qir-profile.ll
+```
+
+Default behaviour of the tool is that it always validates the generated profile. This behaviour can be disabled by
+
+```language
+qat -p profile.yaml --generate --no-validate qir.ll > qir-profile.ll
+```
+
+# Profile Specification
+
+Every profile is specified through a YAML file which defines an object at the top-level. This object must contain the fields `name` and `displayName`:
+
+```yaml
+name: profile-name
+displayName: Profile Name
+# ...
+```
+
+Additionally, top level also contains the fields `version` and `mode`. The version refers to the QIR version which forms the basis for the specification and `mode` explains how the profile is defined.
+The final two top level fields are lists named `specification` and `generation`. These contains the specification and generation producedure, respectively.
+
+## Specification
+
+The default `mode` of specification is by `feature` which means that specification describes the feature set available. Alternatively, one can specify a profile by `limitation`. As an example profile for
```

Comment:
I can imagine profiles where you have a mix of `feature` and `limitation`. For instance, I could imagine specifying the list of quantum instructions by `feature` but wanting to rule out specific LLVM instructions by `limitation`. Might it be more flexible to get rid of `mode` and instead have top-level `feature` and `limitation` tags?

Reply:
Yes, I agree with this. I was thinking exactly the same. I was thinking that this could be a later extension where we would have `mixed` and that we would start with `feature` and `limitation`. It is clear though that in many case it will be beneficial to for instance say "No floating point operations, all integer operation except for division and multiplication" instead either specifying all support or all not supported.